### DISCUSSION TITLE
`snforge_std` new modules ideas

### DISCUSSION THREAD

**Discussion by war-in** [foundry-rs/starknet-foundry#991] [Ideas]
While working on #904 I walked through some Cairo projects and checked what could be improved.

## Checked projects
- [satoru](https://github.com/keep-starknet-strange/satoru/tree/main)
- [nogame](https://github.com/ametel01/nogame-starknet/tree/main)
- [ark-lane](https://github.com/ArkProjectNFTs/ark-lane/tree/dev)
- and a few smaller ones

## `Cairo` issues
- some projects use self-made `time` modules for handling dates
- there is no all-in-one math library for Cairo (correct me if I'm wrong)
- extending arrays requires users to build loops (it could be included in corelib)
- unwrap does not return an unpacked error message

## `SN-Foundry` issues
- while deploying, users always have to instantiate a dispatcher (maybe it could be done inside the `deploy` cheatcode?)
- there is no documentation in the cheatcodes code (it would be easier to jump through the code and read the description than opening the web documentation)

## Features available in `Foundry`

There are a few features available in `Foundry` that `Starknet-Foundry` doesn't have:
- assertions module, providing extended assert functions
- errors module, containing  improved common errors
- logging module, providing advanced logging functionalities

## Summary
I'm opening this discussion to get feedback from the community and developers.

Should we create independent libraries with some of those features or add them to the `snforge_std`?
Also, it would be nice to get your opinion on what is most important for you.

**Comment by sgc-code** [comment]
Hi, here are some issue we are facing when testing. Mostly are not specific to foundry and they also apply to cairo testing in general, But it would be great if those could be solved somehow:

- it's not closely simulating the real network: for instance catching errors when calling another contract is not working on real networks but works on foundry (or any cairo test). Ideally the tests behave the same as the real network
- accurate profiling, cairo tests only return the gas used (and that’s not accurate as it’s skipping the account validation). Devnet on the other hand gives more details like the steps, the built-ins, the storage slots used…
- whitelisted functions are not taken into account in cairo tests
- not able to accurately test transactions: calling the validate method, the fees...
- foundry names are difficult to read and hard to remember: prank, roll, warp…, could be set_caller_address, set_block_number, set_block_timestamp
